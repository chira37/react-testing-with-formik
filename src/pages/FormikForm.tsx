import React, { useState } from "react";
import ReactDOM from "react-dom";
import { useFormik } from "formik";
import * as yup from "yup";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import { Box, Container } from "@material-ui/core";
import { post } from "../../src/utils/api";
interface IFormikForm {
    title: string;
}

const validationSchema = yup.object({
    email: yup.string().email().required(),
    username: yup.string().required(),
    mobile: yup.string().required(),
    password: yup.string().min(8).required(),
});

interface IUser {
    email: string;
    username: string;
    mobile: string;
}

const FormikForm: React.FC<IFormikForm> = ({ title }) => {
    const [user, setUser] = useState<IUser | null>(null);

    const formik = useFormik({
        initialValues: {
            email: "",
            username: "",
            mobile: "",
            password: "",
        },
        validationSchema: validationSchema,
        onSubmit: (values) => {
            post(values).then((res) => {
                setUser(res);
                console.log(res);
            });
        },
    });

    return (
        <Container maxWidth="sm">
            <Box data-testid="form_title" fontSize="h4.fontSize" pt={10} pb={10}>
                {title}
            </Box>
            <form onSubmit={formik.handleSubmit}  data-testid="form">
                <TextField
                    style={{ marginBottom: 25 }}
                    fullWidth
                    id="email"
                    name="email"
                    label="Email"
                    variant="outlined"
                    value={formik.values.email}
                    onChange={formik.handleChange}
                    error={formik.touched.email && Boolean(formik.errors.email)}
                    helperText={formik.touched.email && formik.errors.email}
                />

                <TextField
                    style={{ marginBottom: 25 }}
                    fullWidth
                    id="username"
                    name="username"
                    label="Username"
                    variant="outlined"
                    type="text"
                    value={formik.values.username}
                    onChange={formik.handleChange}
                    error={formik.touched.username && Boolean(formik.errors.username)}
                    helperText={formik.touched.username && formik.errors.username}
                />

                <TextField
                    style={{ marginBottom: 25 }}
                    fullWidth
                    id="mobile"
                    name="mobile"
                    label="Mobile"
                    variant="outlined"
                    value={formik.values.mobile}
                    onChange={formik.handleChange}
                    error={formik.touched.mobile && Boolean(formik.errors.mobile)}
                    helperText={formik.touched.mobile && formik.errors.mobile}
                />
                <TextField
                    style={{ marginBottom: 25 }}
                    fullWidth
                    id="password"
                    name="password"
                    label="Password"
                    type="password"
                    variant="outlined"
                    value={formik.values.password}
                    onChange={formik.handleChange}
                    error={formik.touched.password && Boolean(formik.errors.password)}
                    helperText={formik.touched.password && formik.errors.password}
                />
                <Button data-testid="submit_button" color="primary" variant="contained" fullWidth type="submit">
                    Submit
                </Button>
            </form>
        </Container>
    );
};

export default FormikForm;
