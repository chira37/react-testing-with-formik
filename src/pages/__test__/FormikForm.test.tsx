import React from "react";
import ReactDOM from "react-dom";
import { render, cleanup, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import "@testing-library/jest-dom";
import renderer from "react-test-renderer";

import FormikForm from "../FormikForm";
import { act } from "react-dom/test-utils";

afterEach(cleanup);

it("render without crash", () => {
    const div = document.createElement("div");
    ReactDOM.render(<FormikForm title="TEST" />, div);
});

it("render FormikForm correctly", () => {
    const { getByTestId } = render(<FormikForm title="TESTS" />);
    expect(getByTestId("form_title")).toHaveTextContent("TEST");
});

it("match snapshot", () => {
    const tree = renderer.create(<FormikForm title="TESTS"/>).toJSON();
    expect(tree).toMatchSnapshot();
});

test("check input validation errors", async () => {
    const { getByTestId, queryByText, debug } = render(<FormikForm title="TESTS" />);
    const submitButton = getByTestId("submit_button");

    await act(async () => {
        userEvent.click(submitButton);
    });

    expect(queryByText("email is a required field")).not.toBeNull();
    expect(queryByText("username is a required field")).not.toBeNull();
    expect(queryByText("mobile is a required field")).not.toBeNull();
    expect(queryByText("password is a required field")).not.toBeNull();
    debug();
});

test("check input validation success", async () => {
    const { getByTestId, queryAllByText, getByLabelText, debug } = render(<FormikForm title="TESTS" />);
    const submitButton = getByTestId("submit_button");

    await act(async () => {
        fireEvent.change(getByLabelText("Email"), { target: { value: "test@gmail.com" } });
        fireEvent.change(getByLabelText("Username"), { target: { value: "test" } });
        fireEvent.change(getByLabelText("Mobile"), { target: { value: "12345678891" } });
        fireEvent.change(getByLabelText("Password"), { target: { value: "121232142143" } });
    });

    await act(async () => {
        fireEvent.click(submitButton);
    });

    const validationErrors = queryAllByText("required", { exact: false });
    expect(validationErrors).toEqual([]);
    //debug();
});
