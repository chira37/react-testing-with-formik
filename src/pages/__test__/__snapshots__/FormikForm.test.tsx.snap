// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthSm"
>
  <div
    className="MuiBox-root MuiBox-root-7"
    data-testid="form_title"
  >
    TESTS
  </div>
  <form
    data-testid="form"
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style={
        Object {
          "marginBottom": 25,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="email"
        id="email-label"
      >
        Email
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="email"
          name="email"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-4"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style={
        Object {
          "marginBottom": 25,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="username"
        id="username-label"
      >
        Username
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="username"
          name="username"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-4"
          >
            <span>
              Username
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style={
        Object {
          "marginBottom": 25,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="mobile"
        id="mobile-label"
      >
        Mobile
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="mobile"
          name="mobile"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-4"
          >
            <span>
              Mobile
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style={
        Object {
          "marginBottom": 25,
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="password"
        id="password-label"
      >
        Password
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="password"
          name="password"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="password"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-4"
          >
            <span>
              Password
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
      data-testid="submit_button"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Submit
      </span>
    </button>
  </form>
</div>
`;
